src/beliefpropagation.jl

```diff
@@ -1,41 +1,67 @@
-function construct_initial_mts(
+function message_tensors(
   tn::ITensorNetwork, nvertices_per_partition::Integer; partition_kwargs=(;), kwargs...
 )
-  return construct_initial_mts(
-    tn, partition(tn; nvertices_per_partition, partition_kwargs...); kwargs...
+  return message_tensors(
+    partition(tn; nvertices_per_partition, partition_kwargs...); kwargs...
   )
 end
 
-function construct_initial_mts(
-  tn::ITensorNetwork, subgraphs::DataGraph; init=(I...) -> @compat allequal(I) ? 1 : 0
+function message_tensors(
+  subgraphs::DataGraph; contract_kwargs=(;), init=(I...) -> allequal(I) ? 1 : 0
 )
-  # TODO: This is dropping the vertex data for some reason.
-  # mts = DataGraph{vertextype(subgraphs),vertex_data_type(subgraphs),ITensor}(subgraphs)
-  mts = DataGraph{vertextype(subgraphs),vertex_data_type(subgraphs),ITensor}(
+  mts = DataGraph{vertextype(subgraphs),vertex_data_type(subgraphs),ITensorNetwork}(
     directed_graph(underlying_graph(subgraphs))
   )
   for v in vertices(mts)
     mts[v] = subgraphs[v]
   end
   for subgraph in vertices(subgraphs)
-    tns_to_contract = ITensor[]
     for subgraph_neighbor in neighbors(subgraphs, subgraph)
-      edge_inds = Index[]
-      for vertex in vertices(subgraphs[subgraph])
-        psiv = tn[vertex]
-        for e in [edgetype(tn)(vertex => neighbor) for neighbor in neighbors(tn, vertex)]
-          if (find_subgraph(dst(e), subgraphs) == subgraph_neighbor)
-            append!(edge_inds, commoninds(tn, e))
-          end
-        end
-      end
-      mt = normalize!(
-        itensor(
-          [init(Tuple(I)...) for I in CartesianIndices(tuple(dim.(edge_inds)...))],
-          edge_inds,
+      boundary_vertices_s = setdiff(
+        vertices(subgraphs[subgraph]),
+        unique(
+          flatten([
+            neighbors(subgraphs[subgraph_neighbor], vn) for
+            vn in vertices(subgraphs[subgraph_neighbor])
+          ]),
+        ),
+      )
+      boundary_vertices_sn = setdiff(
+        vertices(subgraphs[subgraph_neighbor]),
+        unique(
+          flatten([
+            neighbors(subgraphs[subgraph], vn) for vn in vertices(subgraphs[subgraph])
+          ]),
         ),
       )
-      mts[subgraph => subgraph_neighbor] = mt
+      boundary_tensors = ITensor[]
+      for v in boundary_vertices_s
+        inds_boundary = flatten(
+          unique([inds(subgraphs[subgraph_neighbor][vn]) for vn in boundary_vertices_sn])
+        )
+        inds_internal = flatten(
+          unique([inds(subgraphs[subgraph][v]) for v in setdiff(boundary_vertices_s, [v])])
+        )
+        new_inds = intersect(
+          inds(subgraphs[subgraph][v]), vcat(inds_boundary, inds_internal)
+        )
+        push!(
+          boundary_tensors,
+          itensor(
+            [init(Tuple(I)...) for I in CartesianIndices(tuple(dim.(new_inds)...))],
+            new_inds,
+          ),
+        )
+      end
```

Comment:
So there is a bit of a "secret" feature in `partition` where it stores the boundary edges of the original partitioned network as the edge metadata of the `DataGraph` representing the partitioned network:
```julia
julia> using Metis

julia> using ITensorNetworks

julia> tn = ITensorNetwork(named_grid((4, 4)); link_space=2);

julia> tnp = partition(tn; nvertices_per_partition=4);

julia> vertices(tnp)
4-element Vector{Int64}:
 1
 2
 3
 4

julia> vertices(tnp[1])
4-element Vector{Tuple{Int64, Int64}}:
 (1, 1)
 (2, 1)
 (1, 2)
 (2, 2)

julia> vertices(tnp[2])
4-element Vector{Tuple{Int64, Int64}}:
 (3, 1)
 (4, 1)
 (3, 2)
 (4, 2)

julia> tnp[1 => 2][:edges]
2-element Vector{NamedGraphs.NamedEdge{Tuple{Int64, Int64}}}:
 (2, 1) => (3, 1)
 (2, 2) => (3, 2)
```
Is that helpful here?

You can see the logic here for extracting those edges here: https://github.com/mtfishman/ITensorNetworks.jl/blob/741a28f8c3c31fd0161816ae1c7fcc379e001cea/src/partition.jl#L235-L257

Reply:
Yeah I think that might be useful actually and help simplify that section if I can easily extract the vertices on the `subgraph` side of the `subgraph -> subgraph_neighbor` partition. I would still need to prune the unnecessary indices off of the tensors on those vertices.

Reply:
It sounds like there are some functions we could make for `ITensorNetwork`s or partitioned `ITensorNetwork`s that would make implementing this kind of function easier.

Ultimately, given two neighboring subgraph `ITensorNetwork`s (say `tnp[1]` and `tnp[2]` above), don't you just need the common indices between those two networks to make the message tensors?

It sounds like if there was a function `commoninds(tn1::ITensorNetwork, tn2::ITensorNetwork)` that output a `Vector{Index}` with the indices shared between two networks, then we could do something like the following:
```julia
function message_tensors(subgraphs::DataGraph; maxdim=1)
  mts = DataGraph{<:Any,<:Any,ITensorNetwork}(directed_graph(underlying_graph(subgraphs)))
  for e in edges(subgraphs)
    inds_e = commoninds(subgraphs[src(e)], subgraphs[dst(e)])
    indsnetwork_e = IndsNetwork(
      path_graph(length(inds_e));
      site_space=Dictionary(eachindex(inds_e), inds_e),
      link_space=maxdim,
    )
    mts[e] = randomITensorNetwork(indsnetwork_e)
    mts[reverse(e)] = dag(mts[e])
  end
  return mts
end
```

Reply:
In principle yes. The only caveat here is that this doesn't account for the internal structure of the boundary tensornetwork? For example if you want to initialise an MPS as the message tensor then you want the ordering of the boundary indices to reflect the order of the tensors of the boundary (say like to match the column of a PEPS) and to not be arbitrary?

Reply:
Hence why I am currently keeping track of the `internal_inds` and constructing a tensor network as the message tensor which has both the correct internal and boundary inds?

Reply:
Although that function above is useful and could be used to make things more streamlined. So should I add it to the PR?

Reply:
Ah, I didn't realize that's what your function is doing. That seems pretty subtle to do in general, how are you determining the linear ordering of the MPS?

It still seems like you could use the function above, but then have a separate step that determines the linear order of `inds_e`.

That seems like a good reason to just use bond dimension one MPS message tensors to start with, since then the linear ordering doesn't matter. And then Linjian's `"density_matrix"` algorithm will determine a good ordering of the MPS message tensors after the first update step (so it sounds like you are essentially reproducing some of the work Linjian does for determining the linear ordering). It just seems like a lot of work and code for what could be a very simple function, if we simplified our goal a little bit.

Reply:
Update: We've decided to just handle product state (rank 1) message tensors for now, so something like:
```julia
function message_tensors(subgraphs::DataGraph; itensor_constructor=randomITensor)
  mts = DataGraph{<:Any,<:Any,ITensorNetwork}(directed_graph(underlying_graph(subgraphs)))
  for e in edges(subgraphs)
    inds_e = commoninds(subgraphs[src(e)], subgraphs[dst(e)])
    mts[e] = ITensorNetwork(map(itensor_constructor, inds_e))
    mts[reverse(e)] = dag(mts[e])
  end
  return mts
end
```
where we are offering some choice about how to construct the ITensors making up the rank 1 message tensor networks with the `itensor_constructor` keyword argument.

Reply:
Okay great, this is what I have now gone for with initialisation -> with the default being `delta()` for `itensor_constructor`.